src/tensors/levels/mutex_levels.jl

```diff
@@ -0,0 +1,272 @@
+
+    push_preamble!(ctx, quote
+        $atomics = $(lvl.locks)
+        $(lvl.locks) = $moveto($(lvl.locks), $(ctx(arch)))
+    end)
+    push_epilogue!(ctx, quote
+        $(lvl.locks) = $atomics
+    end)
+    virtual_moveto_level(ctx, lvl.lvl, arch)
+end
+
+function instantiate(ctx, fbr::VirtualSubFiber{VirtualMutexLevel}, mode::Reader)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    instantiate(ctx, VirtualSubFiber(lvl.lvl, pos), mode)
+end
+
+function unfurl(ctx, fbr::VirtualSubFiber{VirtualMutexLevel}, ext, mode::Reader, proto)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    unfurl(ctx, VirtualSubFiber(lvl.lvl, pos), ext, mode, proto)
+end
+
+function unfurl(ctx, fbr::VirtualSubFiber{VirtualMutexLevel}, ext, mode::Updater, proto)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    sym = freshen(ctx, lvl.ex, :after_atomic_lvl)
+    atomicData = freshen(ctx, lvl.ex, :atomicArraysAcc)
+    lockVal = freshen(ctx, lvl.ex, :lockVal)
+    dev = lower(ctx, virtual_get_device(ctx.code.task), DefaultStyle())
+    push_preamble!(ctx, quote
+        $atomicData =  Finch.get_lock($dev, $(lvl.locks), $(ctx(pos)), eltype($(lvl.AVal)))
+        $lockVal = Finch.aquire_lock!($dev, $atomicData)
+    end)
+    res = unfurl(ctx, VirtualSubFiber(lvl.lvl, pos), ext, mode, proto)
+    push_epilogue!(ctx, quote
+        Finch.release_lock!($dev, $atomicData)
+    end)
+    return res
+end
+
+function unfurl(ctx, fbr::VirtualHollowSubFiber{VirtualMutexLevel}, ext, mode::Updater, proto)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    sym = freshen(ctx, lvl.ex, :after_atomic_lvl)
+    atomicData = freshen(ctx, lvl.ex, :atomicArraysAcc)
+    lockVal = freshen(ctx, lvl.ex, :lockVal)
+    dev = lower(ctx, virtual_get_device(ctx.code.task), DefaultStyle())
+    push_preamble!(ctx, quote
+        $atomicData =  Finch.get_lock($dev, $(lvl.locks), $(ctx(pos)), eltype($(lvl.AVal)))
+        $lockVal = Finch.aquire_lock!($dev, $atomicData)
+    end)
+    res = unfurl(ctx, VirtualHollowSubFiber(lvl.lvl, pos, fbr.dirty), ext, mode, proto)
+    push_epilogue!(ctx, quote
+        Finch.release_lock!($dev, $atomicData)
+    end)
+    return res
+end
+
+function lower_assign(ctx, fbr::VirtualSubFiber{VirtualMutexLevel}, mode::Updater, op, rhs)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    sym = freshen(ctx, lvl.ex, :after_atomic_lvl)
+    atomicData = freshen(ctx, lvl.ex, :atomicArraysAcc)
+    lockVal = freshen(ctx, lvl.ex, :lockVal)
+    dev = lower(ctx, virtual_get_device(ctx.code.task), DefaultStyle())
+    push_preamble!(ctx, quote
+        $atomicData =  Finch.get_lock($dev, $(lvl.locks), $(ctx(pos)), eltype($(lvl.AVal)))
+        $lockVal = Finch.aquire_lock!($dev, $atomicData)
+    end)
+    res = lower_assign(ctx, VirtualSubFiber(lvl.lvl, pos), mode, op, rhs)
+    push_epilogue!(ctx, quote
+        Finch.release_lock!($dev, $atomicData)
+    end)
+    return res
+end
+
+function lower_assign(ctx, fbr::VirtualHollowSubFiber{VirtualMutexLevel}, mode::Updater, op, rhs)
```

Comment:
one thing that I do like about this function is now we could add a way to do an atomicOp via a sort of "HollowAtomic Level", which translates certain ops to atomic operations depending on what the arch supports.

Reply:
Yes, that was the intention, currently that stuff lives in architecture.jl